### ISSUE TITLE
use_azure_cli_auth and shared image gallery

### ISSUE THREAD

**Issue by yellowhat** [hashicorp/packer#10429]
Hi,
I would like to use the `use_azure_cli_auth` option against a Shared Image Gallery, with packer 1.6.6.

I am using the following example to replicate the issue:
```json
{
  "builders": [{
    "type": "azure-arm",
    "use_azure_cli_auth": "True",

    "os_type": "Linux",
    "image_publisher": "OpenLogic",
    "image_offer": "CentOS",
    "image_sku": "7.7",

    "location": "West Europe",
    "vm_size": "Standard_DS2_v2",

    "shared_image_gallery_destination": {
        "image_name":     "ccnode",
        "image_version":  "0.1.0",
        "resource_group": "zzzz_image",
        "gallery_name":   "zsig",
        "replication_regions": ["West Europe"]
    },
    "managed_image_name":                 "ccnode-0.1.0",
    "managed_image_resource_group_name":  "zzzz_image",
    "managed_image_storage_account_type": "Premium_LRS"
  }],
  "provisioners": [{
    "execute_command": "chmod +x {{ .Path }}; {{ .Vars }} sudo -E sh '{{ .Path }}'",
    "inline": [
      "date > /data",
      "/usr/sbin/waagent -force -deprovision+user && export HISTSIZE=0 && sync"
    ],
    "inline_shebang": "/bin/sh -x",
    "type": "shell"
  }]
}
```

Unfortunately I get the following error:
```
$ PACKER_LOG=1 packer build packer.json
==> azure-arm: Getting tokens using Azure CLI
==> azure-arm: Getting tokens using Azure CLI
    azure-arm: Creating Azure Resource Manager (ARM) client ...
2020/12/29 16:01:30 packer-builder-azure-arm plugin: Azure request method="GET" request="https://management.azure.com/subscriptions/sub_id/resourcegroups/zzzz_image?api-version=2018-02-01" body=""
2020/12/29 16:01:30 packer-builder-azure-arm plugin: Azure response x-ms-request-id="fb1537fc-b400-44b5-9c2c-1036065eaa29" body="{\"id\":\"/subscriptions/sub_id/resourceGroups/zzzz_image\",\"name\":\"zzzz_image\",\"location\":\"westeurope\",\"properties\":{\"provisioningState\":\"Succeeded\"}}" status="200 OK" method="GET" request="https://management.azure.com/subscriptions/sub_id/resourcegroups/zzzz_image?api-version=2018-02-01"
2020/12/29 16:01:30 packer-builder-azure-arm plugin: Azure response status="200 OK" method="GET" request="https://management.azure.com/subscriptions/sub_id/resourcegroups/zzzz_image?api-version=2018-02-01" x-ms-request-id="fb1537fc-b400-44b5-9c2c-1036065eaa29" body="{\"id\":\"/subscriptions/sub_id/resourceGroups/zzzz_image\",\"name\":\"zzzz_image\",\"location\":\"westeurope\",\"properties\":{\"provisioningState\":\"Succeeded\"}}"
2020/12/29 16:01:30 packer-builder-azure-arm plugin: Azure request method="GET" request="https://management.azure.com/subscriptions/sub_id/resourceGroups/zzzz_image/providers/Microsoft.Compute/images/ccnode-0.1.0?api-version=2018-04-01" body=""
2020/12/29 16:01:31 packer-builder-azure-arm plugin: Azure response status="404 Not Found" method="GET" request="https://management.azure.com/subscriptions/sub_id/resourceGroups/zzzz_image/providers/Microsoft.Compute/images/ccnode-0.1.0?api-version=2018-04-01" x-ms-request-id="57271a83-594b-4a96-b4fc-0e65d34c9ac3" body="{\"error\":{\"code\":\"ResourceNotFound\",\"message\":\"The Resource 'Microsoft.Compute/images/ccnode-0.1.0' under resource group 'zzzz_image' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix\"}}"
2020/12/29 16:01:31 packer-builder-azure-arm plugin: Azure response x-ms-request-id="57271a83-594b-4a96-b4fc-0e65d34c9ac3" body="{\"error\":{\"code\":\"ResourceNotFound\",\"message\":\"The Resource 'Microsoft.Compute/images/ccnode-0.1.0' under resource group 'zzzz_image' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix\"}}" status="404 Not Found" method="GET" request="https://management.azure.com/subscriptions/sub_id/resourceGroups/zzzz_image/providers/Microsoft.Compute/images/ccnode-0.1.0?api-version=2018-04-01"
==> azure-arm: WARNING: Zone resiliency may not be supported in West Europe, checkout the docs at https://docs.microsoft.com/en-us/azure/availability-zones/
2020/12/29 16:01:31 packer-builder-azure-arm plugin: Azure request method="GET" request="https://management.azure.com/subscriptions//resourceGroups/zzzz_image/providers/Microsoft.Compute/galleries/zsig/images/ccnode?api-version=2019-03-01" body=""
2020/12/29 16:01:31 packer-builder-azure-arm plugin: Azure response status="404 Not Found" method="GET" request="https://management.azure.com/subscriptions//resourceGroups/zzzz_image/providers/Microsoft.Compute/galleries/zsig/images/ccnode?api-version=2019-03-01" x-ms-request-id="6c56b808-7728-41b8-b8e0-5ab3be0521cf" body="{\"error\":{\"code\":\"SubscriptionNotFound\",\"message\":\"The subscription 'resourceGroups' could not be found.\"}}"
2020/12/29 16:01:31 packer-builder-azure-arm plugin: Azure response x-ms-request-id="6c56b808-7728-41b8-b8e0-5ab3be0521cf" body="{\"error\":{\"code\":\"SubscriptionNotFound\",\"message\":\"The subscription 'resourceGroups' could not be found.\"}}" status="404 Not Found" method="GET" request="https://management.azure.com/subscriptions//resourceGroups/zzzz_image/providers/Microsoft.Compute/galleries/zsig/images/ccnode?api-version=2019-03-01"
2020/12/29 16:01:31 [INFO] (telemetry) ending azure-arm
==> Wait completed after 1 second 366 milliseconds
2020/12/29 16:01:31 machine readable: error-count []string{"1"}
==> Some builds didn't complete successfully and had errors:
2020/12/29 16:01:31 machine readable: azure-arm,error []string{"the Shared Gallery Image to which to publish the managed image version to does not exist in the resource group zzzz_image"}
==> Builds finished but no artifacts were created.
Build 'azure-arm' errored after 1 second 366 milliseconds: the Shared Gallery Image to which to publish the managed image version to does not exist in the resource group zzzz_image

==> Wait completed after 1 second 366 milliseconds

==> Some builds didn't complete successfully and had errors:
--> azure-arm: the Shared Gallery Image to which to publish the managed image version to does not exist in the resource group zzzz_image
```

Instead if I use:
```json
{
 "variables": {
        "tenant_id": "tenant_id",
        "subscription_id": "subscription_id",
        "client_id": "client_id",
        "client_secret": "client_secret"
  },
  "builders": [{
    "type": "azure-arm",
    "client_id": "{{user `client_id`}}",
    "client_secret": "{{user `client_secret`}}",
    "subscription_id": "{{user `subscription_id`}}",
    "tenant_id": "{{user `tenant_id`}}",

    "os_type": "Linux",
    "image_publisher": "OpenLogic",
    "image_offer": "CentOS",
    "image_sku": "7.7",

    "location": "West Europe",
    "vm_size": "Standard_DS2_v2",

    "shared_image_gallery_destination": {
        "image_name":     "ccnode",
        "image_version":  "0.1.0",
        "resource_group": "zzzz_image",
        "gallery_name":   "zsig",
        "replication_regions": ["West Europe"]
    },
    "managed_image_name":                 "ccnode-0.1.0",
    "managed_image_resource_group_name":  "zzzz_image",
    "managed_image_storage_account_type": "Premium_LRS"
  }],
  "provisioners": [{
    "execute_command": "chmod +x {{ .Path }}; {{ .Vars }} sudo -E sh '{{ .Path }}'",
    "inline": [
      "date > /data",
      "/usr/sbin/waagent -force -deprovision+user && export HISTSIZE=0 && sync"
    ],
    "inline_shebang": "/bin/sh -x",
    "type": "shell"
  }]
}
```

It works as expected.

Seems that with the `use_azure_cli_auth` it is expecting the image version already present instead of creating a new one.

Thanks

**Comment by yellowhat** [comment]
Is there something else I can do to move this forward?

Thanks

**Comment by SwampDragons** [comment]
Hi, I know waiting on issues that are blockers can be frustrating. I'm sorry to say it but Packer has a very small team of core maintainers, and we're stretched pretty thin. I'm afraid you're going to have to just be patient until we find the time to investigate. You are also welcome to try to find a workaround, or investigate and open a PR yourself.

**Comment by zmne** [comment]
@yellowhat I'm using HCL2 instead of JSON, and it's going through just fine. Maybe give that a shot?

**Comment by yellowhat** [comment]
I have tried the following:
```
locals { 
    timestamp = regex_replace(timestamp(), "[- TZ:]", "")
}

source "azure-arm" "autogenerated_1" {
  image_offer                        = "CentOS"
  image_publisher                    = "OpenLogic"
  image_sku                          = "7_7-gen2"
  location                           = "West Europe"
  managed_image_name                 = "vmimage1-0.1.0"
  managed_image_resource_group_name  = "zz0image0zz"
  managed_image_storage_account_type = "Premium_LRS"
  os_type                            = "Linux"
  shared_image_gallery_destination {
    gallery_name        = "zzz0sig0zzz"
    image_name          = "vmimage1"
    image_version       = "0.1.0"
    replication_regions = ["West Europe"]
    resource_group      = "zz0image0zz"
  }
  temp_compute_name        = "vm{{ replace_all \"-\" \"\" uuid }}"
  temp_resource_group_name = "packercli-{{ strftime \"%Y%m%d_%H%M%S\" }}"
  use_azure_cli_auth       = true
  vm_size                  = "Standard_DS2_v2"
}

build {
  sources = ["source.azure-arm.autogenerated_1"]
  provisioner "shell" {
    execute_command = "chmod +x {{ .Path }}; {{ .Vars }} sudo -E sh '{{ .Path }}'"
    inline          = [
        "date > /data",
        "/usr/sbin/waagent -force -deprovision+user && export HISTSIZE=0 && sync"
    ]
    inline_shebang  = "/bin/sh -x"
  }
}
```

Still getting the error:
```
$ packer build packer_cli.pkr.hcl
azure-arm.autogenerated_1: output will be in this color.

==> azure-arm.autogenerated_1: Running builder ...
==> azure-arm.autogenerated_1: Getting tokens using Azure CLI
==> azure-arm.autogenerated_1: Getting tokens using Azure CLI
    azure-arm.autogenerated_1: Creating Azure Resource Manager (ARM) client ...
==> azure-arm.autogenerated_1: WARNING: Zone resiliency may not be supported in West Europe, checkout the docs at https://docs.microsoft.com/en-us/azure/availability-zones/
Build 'azure-arm.autogenerated_1' errored after 1 second 258 milliseconds: the Shared Gallery Image to which to publish the managed image version to does not exist in the resource group zz0image0zz

==> Wait completed after 1 second 258 milliseconds

==> Some builds didn't complete successfully and had errors:
--> azure-arm.autogenerated_1: the Shared Gallery Image to which to publish the managed image version to does not exist in the resource group zz0image0zz

==> Builds finished but no artifacts were created.
```

Am I doing something different?

**Comment by zmne** [comment]
@yellowhat By reading the error message, this looks more related to Azure issue then Packer itself. Do `az account show`, does this return the subscription the resource group is created in (by default, thats the subscription it will try to communicate with) ? If not, within your `shared_image_gallery_destination` block, add `subscription` pointing to the right id that has that resource group within.

**Comment by animatco** [comment]
I got the same message the 1st time I tried. The answer is pretty simple
though not documented in the Azure-Arm builder docs. You will need to
pre-stage the Shared Gallery definition file. How I did it is as follows
using the Azure CLI:

az sig image-definition create --resource-group "ADT-Packer-Images-rg"
--gallery-name "AnsiblePackerImages" --gallery-image-definition
"WindowsServer2019-Ansible-Packer" --publisher "ADT" --offer
Windowserver --sku 2019 --os-type windows  

mateusz wrote on 1/26/2021 10:38 AM:
